docs/content/_index.md

```diff
@@ -1,608 +1,1007 @@
+input.networks[j].public
+```
+```live:example/iter/join:output
+```
+
+> If you split expressions across multiple lines you can omit the `;` operator.
+
+Providing good names for variables can be hard. If you only refer to the
+variable once, you can replace it with the special `_` (wildcard variable)
+operator. Conceptually, each instance of `_` is a unique variable.
+
+```live:example/iter/wildcard:query:merge_down
+input.servers[_].protocols[_] == "http"
+```
+```live:example/iter/wildcard:output
+```
+
+Just like references that refer to non-existent fields or expressions that fail
+to match, if OPA is unable to find any variable assignments that satisfy all of
+the expressions, the result is undefined.
+
+```live:example/iter/undefined:query:merge_down
+some i; input.servers[i].protocols[i] == "ssh"  # there is no assignment of i that satisfies the expression
+```
+```live:example/iter/undefined:output:expect_undefined
+```
+
+### Rules
+
+Rego lets you encapsulate and reuse logic with rules. Rules are just if-then
+logic statements. Rules can either be complete or incremental.
+
+#### Complete Rules
+
+Complete rules are if-then statements that assign a single value to a variable.
+For example:
+
+```live:example/complete/1:module:openable
+package example.rules
+
+any_public_networks = true {
+    net := input.networks[_]
+    net.public
 }
 ```
 
-As a result, any document, base or virtual, can be accessed hierarchically starting from the root data node – either as an identifier:
+Every rule consists of a _head_ and a _body_. In Rego we say the rule head is
+true _if_ the rule body is true for some set of variable assignments. In the
+example above `any_public_networks = true` is the head and `net :=
+input.networks[_]; net.public` is the body.
 
-```ruby
-import data.servers                            # Base document
-import data.opa.examples.violations            # Virtual document
+You can query for the value generated by rules just like any other value:
+
+```live:example/complete/1:query
+any_public_networks
+```
+```live:example/complete/1:output
 ```
 
-or as a URI component in an HTTP request:
+All values generated by rules can be queried via the global `data` variable.
 
-```http
-GET https://example.com/v1/data/servers HTTP/1.1
+```live:example/complete/1/abs:query:merge_down
+data.example.rules.any_public_networks
+```
+```live:example/complete/1/abs:output:merge_down
 ```
 
-```http
-GET https://example.com/v1/data/opa/examples/violations HTTP/1.1
+> 💡 You can query the value of any rule loaded into OPA by referring to it with an
+> absolute path. The path of a rule is always:
+> `data.<package-path>.<rule-name>`.
+
+If you omit the `= <value>` part of the rule head the value defaults to `true`.
+You could rewrite the example above as follows without changing the meaning:
+
+```live:example/complete_elided:module:read_only,openable
+package example.rules
+
+any_public_networks {
```

Comment:
Mostly just brainstorming.. It might be useful to add comments on this similar to what is down at the `Putting It Together` section, so something like:

```
any_public_networks {         # is true if..
    net := input.networks[_]  # some network exists..
    net.public                # and the network is public
}
```
The reason I bring this up is that it might help to kind of push the "right" way to think/reason about these rules early on when learning about them.

Reply:
Done.